### DISCUSSION TITLE
PDF feature

### DISCUSSION THREAD

**Discussion by AtomicNess123** [BurntSushi/ripgrep#1582] [General] [answered]
Does ripgrep search within PDFs?

**Comment by okdana** [comment, marked as answer]
You can use it to do a 'dumb' search of PDFs (the same way you'd search any other binary file) with the `-a` option. If you want to search in a way that semi-respects the actual PDF format, you can use the `--pre` option with something like `pdftotext`. There's an example of that in the [user guide](https://github.com/BurntSushi/ripgrep/blob/master/GUIDE.md#preprocessor)

**Reply by AtomicNess123** [reply]
Thanks. 1) I understand that pdftotext is not included in ripgrep and has to be installed separately? 2) I don't really understand the process to find PDFs (I'm no programmer). The file ".preprocess" contains the following?
```
$ cat processor
#!/bin/sh

case "$1" in
*.pdf)
  # The -s flag ensures that the file is non-empty.
  if [ -s "$1" ]; then
    exec pdftotext - -
  else
    exec cat
  fi
  ;;
*)
  case $(file "$1") in
  *Zstandard*)
    exec pzstd -cdq
    ;;
  *)
    exec cat
    ;;
  esac
  ;;
esac

```

And must be created in the same directory where I want to search for PDFs?
If you know of instructions for dummies, l would appreciate it. Thanks.

**Reply by okdana** [reply]
> I understand that pdftotext is not included in ripgrep and has to be installed separately?

Yes, it's part of the Poppler project. You can install it with `brew install poppler` on macOS, `apt-get install poppler-utils` on Debian/Ubuntu, &c.

> The file ".preprocess" contains the following? ... And must be created in the same directory where I want to search for PDFs?

`preprocess` is the name of the script used in the example, but it can be anything you want. It was located in the same directory as the PDF in the example, but it can be anywhere. You pass the name/location of the script to the `--pre` option when you call `rg`. If it's not in your `PATH` (one of the standard directories where programs like `ls` and `grep` live), you have to explicitly tell `rg` where it is, which is why `./preprocess` is used in the example (`./` is the current directory, which is usually not in `PATH`).

The script can contain anything you want, whatever is useful for your situation, but typically it will look at the file name passed to it (`$1` in the example script) and call different programs depending on what the extension is. If you're going to use `--pre-glob '*.pdf'` to restrict calling the script to just `.pdf` files, you can probably just use the much simpler version shown earlier in that section of the guide (that just does `exec pdftotext - -`).

If you run `rg --help` and look at the sections for `--pre` and `--pre-glob` there is some more information which you should probably look over, but here are the for-dummies instructions (`#` begins a comment, `%` begins a command you should type into the shell):

```
# Install poppler
# (Assuming you're on macOS and have Homebrew; will be something different if not)
% brew install poppler

# Create a directory for your pre-processing script
# (Can be anything, but this puts a .bin in your home directory)
% mkdir -p ~/.bin

# Create the script
% printf '%s\n' '#!/bin/sh' 'exec pdftotext - -' > ~/.bin/preprocess-pdf

# Make the script executable
% chmod a+x ~/.bin/preprocess-pdf

# Use --pre + --pre-glob to tell rg to call the script for .pdf files
% rg --pre-glob '*.pdf' --pre ~/.bin/preprocess-pdf 'my pattern' path/to/pdfs
```

**Reply by AtomicNess123** [reply]
Thanks for the detailed instructions, I followed the instructions and it works!
I was wondering if ripgrep can do searches for several keywords: let's say I want to find documents (PDF) that contain 4 words (e.g., 'apple' 'cake' 'cream' 'oven', to search most likely for recipes including those matches). Is this possible with ripgrep?

**Reply by okdana** [reply]
If you mean you want it to find *all* of those words together (apple AND cake AND cream AND oven), no, it can't really do that right now, though in simpler cases you could use a pattern like `apple.*cake|cake.*apple`.

(Also, ripgrep is designed mainly to search *lines*, not entire documents. You can use the `-U` option to search across multiple lines, but if you're trying to do like full-text document searches it's not really the best tool. Maybe that'll change if/when #1497 is implemented, but idk.)

If you mean you want it to find *any* of those words (apple OR cake OR cream OR oven), it's just like `grep`:

`rg 'apple|cake|cream|oven'`
or
`rg -e apple -e cake -e cream -e oven`

**Reply by AtomicNess123** [reply]
Thanks, I understand. However, I can't understand the grammar

`apple.*cake|cake.*apple.`

Is this searching documents that contain the two words? What if I want to search for three or more words?

The example for any of those words is clear, thanks.

**Reply by AtomicNess123** [reply]
My last question pertains to the "rg --pre ./preprocess 'The Commentz-Walter algorithm' 1995-watson.pdf" syntax. Would it be possible to search in more than one PDF, e.g., in a folder of PDFs? This is not clear. Thanks!

**Reply by okdana** [reply]
> Is this searching documents that contain the two words? What if I want to search for three or more words?

It matches either `apple` followed by `cake` or `cake` followed by `apple` (on the same line, normally). If you want to match more words you have to expand the pattern to include every possible combination, or you can try to use look-around with PCRE, but it obv becomes irritating/infeasible very quickly.

You can read about regular-expression syntax in general at e.g. https://www.regular-expressions.info/quickstart.html, and about the specific syntax that ripgrep uses at https://docs.rs/regex/*/regex/#syntax

> Would it be possible to search in more than one PDF, e.g., in a folder of PDFs?

Just give it the path to a folder instead of a file.

You'll probably want to at least skim through the manual (`rg --help`) and the first few links [here](https://github.com/BurntSushi/ripgrep#documentation-quick-links), especially the guide i linked earlier — they have answers to a lot of basic usage questions like that

**Reply by AtomicNess123** [reply]
Perfect. Thanks for all your help and time.

**Comment by AtomicNess123** [comment]
Sorry to re-open: I just discovered https://github.com/phiresky/ripgrep-all. This function would do what `rg` does without having the need to set up a preprocess file.  With the same functionality?Could you confirm this?

**Reply by BurntSushi** [reply]
I haven't used it. Why not try it and find out?

**Reply by AtomicNess123** [reply]
Yes, it does!
